kernel/src/main/java/org/kframework/parser/inner/kernel/EarleyParser.java

```diff
@@ -0,0 +1,1023 @@
+          predict(S, Q, state, k, data);
+        } else {
+          // state is finished with a production, therefore complete
+          complete(S, Q, state, k, data);
+        }
+        // states ready to process a terminal are added to Q and thus do not appear in S
+      }
+      S.get(k).finish();
+
+      // loop through Q and process each state, scanning it
+      for (EarleyState state : Q) {
+        scan(S, Qprime, state, k, data);
+      }
+
+      // this loop condition is a bit tricky because we want to stop at the exact loop iteration that parsing failed
+      // in order to generate the correct error message. Nominally speaking, parsing succeeds if after all loop
+      // iterations, S[data.words.length] contains the state (productions.get(0), 2, 0). However, if S[k+1] is empty and
+      // Q is empty at this point, then Qprime must also be empty. Thus, if k+1==data.words.length, S[data.words.length]
+      // is empty, which means it's a parse error. And if k+1<data.words.length, then in the next loop iteration,
+      // S[k] and Q will both start out empty, which means that no more states can ever be added at any point. As a
+      // result, it will eventually be a parse error. Thus, the following condition is sufficient to prove that a
+      // parse error will occur, and we can therefore report it now. The condition is not necessary, so we check for
+      // a parse error after the outer loop finishes also.
+      if (S.get(k+1).empty() && Q.empty()) {
+        // no more states to process, so it must be a parse error
+        parseError(data, k);
+      }
+    }
+    if (S.get(data.words.length).empty()) {
+      parseError(data, data.words.length-1);
+    }
+    // finished parsing successfully, so return the final parse forest
+    return Ambiguity.apply(S.get(data.words.length).states.get(0).parseTree().stream().map(list -> list.get(0)).collect(Collectors.toSet()));
+  }
+
+  /**
+   * @param data The {@link ParserMetadata} about the sentence being parsed
+   * @param k The end-index at which a parse error occurred. In other words, the index just prior to the first token that
+   *          did not parse.
+   */
+  private void parseError(ParserMetadata data, int k) {
+    int startLine, startColumn, endLine, endColumn;
+    if (data.words.length == 1) {
+      startLine = data.lines[0];
+      startColumn = data.columns[0];
+      endLine = data.lines[0];
+      endColumn = data.columns[0];
+    } else {
+      Scanner.Token t = data.words[k];
+      startLine = data.lines[t.startLoc];
+      startColumn = data.columns[t.startLoc];
+      endLine = data.lines[t.endLoc];
+      endColumn = data.columns[t.endLoc];
+    }
+    String msg = data.words.length-1 == k ?
+        "Parse error: unexpected end of file" :
+        "Parse error: unexpected token '" + data.words[k].value + "'";
+    if (k != 0) {
+      msg = msg + " following token '" + data.words[k - 1].value + "'.";
```

Comment:
Since Erik added the underline to error messages do we still need this message?

Reply:
Maybe not... I would suggest we make this change in a future PR though.

Reply:
Fair point, minimize the diff.